middlewares/retry.go

```diff
@@ -56,8 +63,34 @@ func (retry *Retry) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
 	}
 }
 
-func isNetworkError(status int) bool {
-	return status == http.StatusBadGateway || status == http.StatusGatewayTimeout
+// netErrorCtxKey is a custom type that is used as key for the context.
+type netErrorCtxKey string
+
+// defaultNetErrCtxKey is the actual key which value is used to record network errors.
+var defaultNetErrCtxKey netErrorCtxKey = "NetErrCtxKey"
+
+// NetErrorRecorder is an interface to record net errors.
+type NetErrorRecorder interface {
+	// Record can be used to signal the retry middleware that an network error happened
+	// and therefore the request should be retried.
+	Record(ctx context.Context)
+}
+
+// NetErrorRecorderFunc is an util to be able to implement the NetErrorRecorder without the need of a struct.
+type NetErrorRecorderFunc func(ctx context.Context)
```

Comment:
The implementation has no need of any field. I originally had a `struct` here without fields, but got a hint in a previous review here that a function would suffice. Therefore I added this type in order to able to implement the `NetErrorRecorder` interface on a function.